### PR TITLE
feat: Add azure blob storage as backend storage

### PR DESCRIPTION
**Pull request by ManuInNZ**
Added ability to have an [Azure Storage Account](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/storage/azblob) as backend storage.

azure-account-name: The globally unique Azure Storage account name
azure-container-name: The container for the blobs
azure-account-key (optional, as it could use the [Azure Default Credentials](https://learn.microsoft.com/en-us/azure/developer/go/azure-sdk-authentication) approach)

Tested and woking but I guess we could improve speed by filtering further on prefixes for the Find.

### CODE REVIEW

File: pkg/storage/azure/resolver.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+		if err != nil {
+			return "", fmt.Errorf("could not get next page: %v", err)
+		}
+		for _, blob := range page.Segment.BlobItems {
+			if *blob.Name == keys {
+				url, err := r.GetSASURL(*blob.Name)
+				if err != nil {
+					return "", fmt.Errorf("could not get SAS URL: %v", err)
+				}
+				fmt.Printf("URL: %s\n", url)
```
Comment (valentindeaconu) [inline pkg/storage/azure/resolver.go:109]: Can you please remove this?

File: pkg/storage/azure/resolver.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		sasQueryParams, err := sas.BlobSignatureValues{
+			Protocol:      sas.ProtocolHTTPS,
+			StartTime:     time.Now().UTC().Add(time.Second * -10),
+			ExpiryTime:    time.Now().UTC().Add(15 * time.Minute),
+			Permissions:   to.Ptr(sas.BlobPermissions{Read: true, List: true}).String(),
+			ContainerName: r.ContainerName,
+			BlobName:      blobName,
+		}.SignWithUserDelegation(udc)
+		if err != nil {
+			log.Fatal(err.Error())
```
Comment (valentindeaconu) [inline pkg/storage/azure/resolver.go:78]: Why is this logging the error instead of returning it?
```suggestion
			return "", err.Error()
```

**Review by valentindeaconu — COMMENTED** [review]
Thank you for this amazing contribution!

Can you please also update the documentation (configuration and readme)?

**Comment by ManuInNZ** [comment]
Sorry I thought I cleaned up. Will put a few words on documentation.

File: pkg/storage/azure/resolver.go (hunk 2 above)
Comment (ManuInNZ) [reply pkg/storage/azure/resolver.go]: yep sorry, fixed

File: pkg/storage/azure/resolver.go (hunk 1 above)
Comment (ManuInNZ) [reply pkg/storage/azure/resolver.go]: Done

**Review by ManuInNZ — COMMENTED** [review]
Voila

**Review by valentindeaconu — APPROVED** [review]
Looks great, thank you!

**Comment by ManuInNZ** [comment]
Arggh just forgot the README updates.



Le mar. 30 avr. 2024 à 23:59, Valentin Deaconu ***@***.***> a
écrit :

**Pull request merged**